src.freebsd/coreutils/uname/uname.1
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd September 14, 2022
.Dt UNAME 1
.Os
.Sh NAME
.Nm uname
.Nd display information about the system
.Sh SYNOPSIS
.Nm
.Op Fl abiKmnoprsUv
.Sh DESCRIPTION
The
.Nm
command writes the name of the operating system implementation to
standard output.
When options are specified, strings representing one or more system
characteristics are written to standard output.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl a
Behave as though the options
.Fl m , n , r , s ,
and
.Fl v
were specified.
.It Fl b
Write the kernel's linker-generated build-id to standard output.
.It Fl i
Write the kernel ident to standard output.
.It Fl K
Write the
.Fx
version of the kernel.
.It Fl m
Write the type of the current hardware platform to standard output.
.Po Xr make 1
uses it to set the
.Va MACHINE
variable.
.Pc
.It Fl n
Write the name of the system to standard output.
.It Fl o
This is a synonym for the
.Fl s
option, for compatibility with other systems.
.It Fl p
Write the type of the machine processor architecture to standard output.
.Po Xr make 1
uses it to set the
.Va MACHINE_ARCH
variable.
.Pc
.It Fl r
Write the current release level of the kernel
to standard output.
.It Fl s
Write the name of the operating system implementation to standard output.
.It Fl U
Write the
.Fx
version of the user environment.
.It Fl v
Write the version level of this release of the operating system
to standard output.
.El
.Pp
If the
.Fl a
flag is specified, or multiple flags are specified, all
output is written on a single line, separated by spaces.
.Pp
The
.Fl K
and
.Fl U
flags are intended to be used for fine grain differentiation of incremental
.Fx
development and user visible changes.
Note that when both of these two options are specified, regardless of their
order, the kernel version would be printed first, followed by the user
environment version.
.Sh ENVIRONMENT
An environment variable composed of the string
.Ev UNAME_
followed by any flag to the
.Nm
utility (except for
.Fl a )
will allow the corresponding data to be set to the contents
of the environment variable.
See
.Xr uname 3
for more information.
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
The hardware platform
.Pq Fl m
can be different from the machine's processor architecture
.Pq Fl p ,
e.g., on 64-bit PowerPC,
.Fl m
would return
.Va ppc64
and
.Fl p
would return
.Va unknown .
.Sh SEE ALSO
.Xr feature_present 3 ,
.Xr getosreldate 3 ,
.Xr sysctl 3 ,
.Xr uname 3 ,
.Xr sysctl 8
.Sh STANDARDS
The
.Nm
command is expected to conform to the
.St -p1003.2
specification.
.Sh HISTORY
The
.Nm
command appeared in PWB UNIX 1.0, however
.Bx 4.4
was the first Berkeley release with the
.Nm
command.
.Pp
The
.Fl K
and
.Fl U
extension flags appeared in
.Fx 10.0 .
The
.Fl b
extension flag appeared in
.Fx 13.0 .


src.freebsd/coreutils/uname/uname.c
/*-
 * SPDX-License-Identifier: BSD-4-Clause
 *
 * Copyright (c) 2002 Juli Mallett.
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <sys/param.h>
#include <sys/utsname.h>
#include <sys/types.h>

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define	MFLAG	0x01
#define	NFLAG	0x02
#define	PFLAG	0x04
#define	RFLAG	0x08
#define	SFLAG	0x10
#define	VFLAG	0x20
#define	IFLAG	0x40
#define	UFLAG	0x80
#define	KFLAG	0x100
#define	BFLAG	0x200

typedef void (*get_t)(void);
static get_t get_buildid, get_ident, get_platform, get_hostname, get_arch,
    get_release, get_sysname, get_kernvers, get_uservers, get_version;

static void native_ident(void);
static void native_platform(void);
static void native_hostname(void);
static void native_arch(void);
static void native_release(void);
static void native_sysname(void);
static void native_version(void);
static void native_kernvers(void);
static void native_uservers(void);
static void native_buildid(void);
static void print_uname(u_int);
static void setup_get(void);
static void usage(void) __dead2;

static char *buildid, *ident, *platform, *hostname, *arch, *release, *sysname,
    *version, *kernvers, *uservers;
static int space;

int
main(int argc, char *argv[])
{
	u_int flags;
	int ch;

	setup_get();
	flags = 0;

	while ((ch = getopt(argc, argv, "abiKmnoprsUv")) != -1)
		switch(ch) {
		case 'a':
			flags |= (MFLAG | NFLAG | RFLAG | SFLAG | VFLAG);
			break;
		case 'b':
			flags |= BFLAG;
			break;
		case 'i':
			flags |= IFLAG;
			break;
		case 'K':
			flags |= KFLAG;
			break;
		case 'm':
			flags |= MFLAG;
			break;
		case 'n':
			flags |= NFLAG;
			break;
		case 'p':
			flags |= PFLAG;
			break;
		case 'r':
			flags |= RFLAG;
			break;
		case 's':
		case 'o':
			flags |= SFLAG;
			break;
		case 'U':
			flags |= UFLAG;
			break;
		case 'v':
			flags |= VFLAG;
			break;
		case '?':
		default:
			usage();
		}

	argc -= optind;
	argv += optind;

	if (argc)
		usage();

	if (!flags)
		flags |= SFLAG;

	print_uname(flags);
	exit(0);
}

#define	CHECK_ENV(opt,var)				\
do {							\
	if ((var = getenv("UNAME_" opt)) == NULL) {	\
		get_##var = native_##var;		\
	} else {					\
		get_##var = (get_t)NULL;		\
	}						\
} while (0)

static void
setup_get(void)
{
	CHECK_ENV("s", sysname);
	CHECK_ENV("n", hostname);
	CHECK_ENV("r", release);
	CHECK_ENV("v", version);
	CHECK_ENV("m", platform);
	CHECK_ENV("p", arch);
	CHECK_ENV("i", ident);
	CHECK_ENV("K", kernvers);
	CHECK_ENV("U", uservers);
	CHECK_ENV("b", buildid);
}

#define	PRINT_FLAG(flags,flag,var)		\
	if ((flags & flag) == flag) {		\
		if (space)			\
			printf(" ");		\
		else				\
			space++;		\
		if (get_##var != NULL)		\
			(*get_##var)();		\
		printf("%s", var);		\
	}

static void
print_uname(u_int flags)
{
	PRINT_FLAG(flags, SFLAG, sysname);
	PRINT_FLAG(flags, NFLAG, hostname);
	PRINT_FLAG(flags, RFLAG, release);
	PRINT_FLAG(flags, VFLAG, version);
	PRINT_FLAG(flags, MFLAG, platform);
	PRINT_FLAG(flags, PFLAG, arch);
	PRINT_FLAG(flags, IFLAG, ident);
	PRINT_FLAG(flags, KFLAG, kernvers);
	PRINT_FLAG(flags, UFLAG, uservers);
	PRINT_FLAG(flags, BFLAG, buildid);
	printf("\n");
}

#define NATIVE_UNAME_GET(var,field) \
static void \
native_##var(void) \
{ \
	static char buf[1024]; \
	struct utsname u; \
 \
	if (uname(&u) != 0) \
		err(1, "uname"); \
\
	snprintf(buf, sizeof(buf), "%s", u.field); \
	var = buf; \
}

NATIVE_UNAME_GET(sysname, sysname)
NATIVE_UNAME_GET(hostname, nodename)
NATIVE_UNAME_GET(release, release)
NATIVE_UNAME_GET(version, version)
NATIVE_UNAME_GET(platform, machine)

#define UNKNOWN_GET(var) \
static void \
native_##var(void) \
{ \
	var = "unknown"; \
}

UNKNOWN_GET(arch)
UNKNOWN_GET(ident)
UNKNOWN_GET(buildid)
UNKNOWN_GET(uservers)
UNKNOWN_GET(kernvers)

static void
usage(void)
{
	fprintf(stderr, "usage: uname [-abiKmnoprsUv]\n");
	exit(1);
}
